### PR TITLE
 bug: Use one mutex per data kind in CURLSH.

### LINKED ISSUE #2711: Deadlock in storage::internal::CurlClient::LockShared()
This was reported by one of our customers, I am opening the bug for tracking.

The customer reports that an application deadlocks, with a single thread calling a function similar to:

```C++
void F(char const* buf, std::size_t len) {
  auto client = gcs::Client::CreatedDefaultClient().value();
  auto stream = client.WriteObject("my-bucket", "my-object");
  stream.write(buf, len);
  stream.Close();
}
```

The stack trace shows the thread blocked on the mutex here:

https://github.com/googleapis/google-cloud-cpp/blob/bcefc9f6322860b546ee5727cb63cdb191592a7c/google/cloud/storage/internal/curl_client.cc#L1148

No other threads are active in the application.

Several instances of the application deadlocked around the same time (different processes, and different machines).

### PR DESCRIPTION
**Pull request by coryan**
I misread the documentation for libcurl and created a single mutex for
all the "kinds" of data in a CURLSH handle. This is incorrect:

https://curl.haxx.se/mail/lib-2019-06/0011.html

I have been unable to reproduce the deadlocks that result from this, so
no additional tests.

I wrote a program to reproduce the bug, without success, but I think the
code and the documentation in that program is useful: it shows how to run
a test on multiple machines use Kubernetes.

This fixes #2711.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2730)

### CODE REVIEW

File: google/cloud/storage/internal/curl_client.h [hunk 1]
```diff
@@ -224,8 +224,13 @@ class CurlClient : public RawClient,
   std::string xml_download_endpoint_;
   std::string iam_endpoint_;
 
+  std::mutex mu_share_;
+  std::mutex mu_dns_;
+  std::mutex mu_ssl_session_;
+  std::mutex mu_connect_;
+  CurlShare share_;
+
   std::mutex mu_;
```
Comment (devbww) [inline google/cloud/storage/internal/curl_client.h]: Now that this looks like it only protects `generator_`, perhaps a new `GUARDED_BY` comment on that member would be illuminating.

Comment (coryan) [reply google/cloud/storage/internal/curl_client.h]: Done.

**Comment by coryan** [comment]
PTAL.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**